m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -143,19 +157,40 @@ private M3Reporter(Builder builder) {
         bucketValueTagKey = builder.histogramBucketName;
         bucketValFmt = String.format("%%.%df", builder.histogramBucketTagPrecision);
 
-        metricQueue = new LinkedBlockingQueue<>(builder.maxQueueSize);
+        queue = new LinkedBlockingQueue<>(builder.maxQueueSize);
+
+        ThreadFactory namedThreadFactory = createThreadFactory();
 
-        executor = builder.executor != null ? builder.executor : Executors.newFixedThreadPool(NUM_PROCESSORS);
+        executorService = builder.executor != null ? builder.executor : Executors.newFixedThreadPool(NUM_PROCESSORS, namedThreadFactory);
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(namedThreadFactory);
 
         clock = Clock.systemUTC();
 
         commonTags = builder.metricTagSet;
 
+        protocolFactory = thriftProtocolFactory;
+
         processorsShutdownLatch = new CountDownLatch(NUM_PROCESSORS);
 
-        processors = new ArrayList<>();
+        collectorEndpointSockedAddresses = builder.endpointSocketAddresses;
+        processors = new Processor[NUM_PROCESSORS];
+
         for (int i = 0; i < NUM_PROCESSORS; ++i) {
-            processors.add(bootProcessor(builder.endpointSocketAddresses));
+            processors[i] = bootProcessor(collectorEndpointSockedAddresses);
+        }
+
+        // Schedule regular heartbeat up-keeping processors up and running
+        scheduledExecutorService.scheduleAtFixedRate(this::heartbeat, 0, HEARTBEAT_PERIOD.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    // NOTE: This method is not concurrent
+    void heartbeat() {
+        synchronized (this) {
+            for (int i = 0; i < processors.length; ++i) {
+                if (processors[i].getState() != ProcessorState.RUNNING) {
+                    processors[i] = bootProcessor(collectorEndpointSockedAddresses);
```

Comment:
Are we guaranteed at this point that the processor has completed shutdown? Looks like the sequence is:

```
            state.set(ProcessorState.SHUTDOWN);

            LOG.warn("Processor shutting down");
            //shutdown
            // Drain queue of any remaining metrics submitted prior to shutdown;
            runNoThrow(this::drainQueue);
            // Flush remaining buffers at last (best effort)
            runNoThrow(this::flushBuffered);

            // Close transport
            transport.close();

            // Count down shutdown latch to notify reporter
            processorsShutdownLatch.countDown();
        }
```

Seems like shutdown of the old processor (queue draining, flushing) could still be going on when a new processor gets rebooted--is that intentional?

Reply:
Yes, those aren't affecting each other in any way

Reply:
They do share the queue though, right? Not that they would necessarily conflict, but we would have two processors concurrently flushing iiuc. I don't know the full context, but would that cause problems? Reason I ask is that there's that big comment about "NUM_PROCESSORS must == 1", so I was wondering whether implicitly breaking that invariant here would cause problems.